examples/maze-solver/maze.py

```diff
@@ -0,0 +1,204 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import dwavebinarycsp as dbc
+from re import match
+
+
+def build_bqm(n_rows, n_cols, start, end, walls, penalty_per_tile=0.5):
+    """Returns a BQM that corresponds to a valid path through a maze. This maze is described by the parameters.
+
+    Specifically, it uses the parameters to build a maze constraint satisfaction problem (CSP). This maze CSP is then
+    converted into the returned BQM.
+
+    Note: If penalty_per_tile is too large, the path will be too heavily penalized and the optimal solution might
+    produce no path at all.
+
+    Args:
+        n_rows: Integer. The number of rows in the maze.
+        n_cols: Integer. The number of cols in the maze.
+        start: String. The location of the starting point of the maze. String follows the format of get_label(..).
+        end: String. The location of the end point of the maze. String follows the format of get_label(..).
+        walls: List of Strings. The list of inner wall locations. Locations follow the format of get_label(..).
+        penalty_per_tile: A number. Penalty for each tile that is included in the path; encourages shorter paths.
+
+    Returns:
+        A dimod.BinaryQuadraticModel
+    """
+    maze = Maze(n_rows, n_cols, start, end, walls)
+    return maze.get_bqm(penalty_per_tile)
+
+
+def get_label(row, col, direction):
+    """Provides a string that follows a standard format for naming constraint variables in Maze.
+    Namely, "<row_index>,<column_index><north_or_west_direction>".
+
+    Args:
+        row: Integer. Index of the row.
+        col: Integer. Index of the column.
+        direction: String in the set {'n', 'w'}. 'n' indicates north and 'w' indicates west.
+    """
+    return "".join([str(row), ",", str(col), direction])
+
+
+def check_label_format(label):
+    """Checks that label conforms with the standard format for naming constraint variables in Maze.
+    Namely, "<row_index>,<column_index><north_or_west_direction>".
+
+    Args:
+        label: String.
+    """
+    assert isinstance(label, str), "Label {} needs to be a string."
+    assert label[-1] in ['n', 'w'], ("Label {} must end with a direction of either 'n' (north) or 'w' "
+                                     "(west). Ex: '2,3n'").format(label)
+
+    row, col = label[:-1].split(",")
+    assert row.isdigit() and int(row) > -1, "Label {} needs to have a non-negative row index.".format(label)
+    assert col.isdigit() and int(col) > -1, "Label {} needs to have a non-negative column index.".format(label)
```

Comment:
What if `label` has a whitespace appended? (assertion fails)
What if `label` has one comma to many? (check fails with `ValueError`)

Generally, asserts should be used for internal sanity check, not for outward propagated data validation. It would be just as good to just assume label is "valid", deconstruct it, and fail with whatever exception is raised during decomposition (if any).

If you stick to asserts, I would rename the function to `assert_label_format_valid`.

If you want to use exceptions to signal invalid label, it would probably be better to use `ValueError`, or even subclass it. Also, better function name in that case would be `validate_label_format`.

More performant version would use booleans to signal validity, and function would be named `is_label_format_valid`.

In any of the cases, I would use regex(es) for parsing.

Reply:
You're right. Regex is a lot cleaner.